### PR TITLE
Задание №2

### LINKED ISSUE #5: Задание №2
Пинг-понг

### PR DESCRIPTION
**Pull request by kot239**
Пинг-понг

### CODE REVIEW

**Comment by TanVD** [comment]
Do you actually need so much style rules for Kotlin in editorconfig? It would also be helpful to add there `end_of_line` and `max_line_length` rules :)

**Comment by TanVD** [comment]
Why do you do `chmod +x gradlew` in CI? Why not to add gradlew with executable permissions under version control?

File: src/main/kotlin/ru/hse/smartUniversity/Main.kt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package ru.hse.smartUniversity
+
+fun main() {
+    println("Very smart")
```
Comment (TanVD) [inline src/main/kotlin/ru/hse/smartUniversity/Main.kt]: What?)

File: src/test/kotlin/ru/hse/smartUniversity/RouteFindingTest.kt [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        fun routes(): List<Arguments> {
+            val mapSingle = UniversityMap()
+            mapSingle.addRoom("101")
+
+            val mapTwoRooms = UniversityMap()
+            mapTwoRooms.addRoom("101")
+            mapTwoRooms.addRoom("102")
+            mapTwoRooms.connect("101", "102")
+
+            val mapLineRooms = UniversityMap()
```
Comment (TanVD) [inline src/test/kotlin/ru/hse/smartUniversity/RouteFindingTest.kt]: Some DSL might not hurt here at all...

File: src/test/kotlin/ru/hse/smartUniversity/RouteFindingTest.kt [hunk 3]
```diff
@@ -0,0 +1,198 @@
+package ru.hse.smartUniversity
+
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.assertThrows
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.Arguments
+import org.junit.jupiter.params.provider.MethodSource
+
+internal class RouteFindingTest {
```
Comment (TanVD) [inline src/test/kotlin/ru/hse/smartUniversity/RouteFindingTest.kt:7]: So, there are basically 4 tests here? It is cool that you are using factories, but in the task I've asked you to write 15 tests...

**Comment by TanVD** [comment]
So... 4.5? Please, add a bit more tests...

**Review by TanVD — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 4]
```diff
@@ -14,13 +15,21 @@ ktlint {
...
 }
 
+compileKotlin { kotlinOptions.jvmTarget = 1.8 }
+
 dependencies {
     testImplementation("org.junit.jupiter:junit-jupiter-engine:5.8.1")
     testImplementation("org.junit.jupiter:junit-jupiter-params:5.8.1")
-    implementation "org.jetbrains.kotlin:kotlin-stdlib"
+    implementation("org.jetbrains.kotlin:kotlin-stdlib")
+    implementation("org.jacoco:org.jacoco.core:0.8.7")
```
Comment (TanVD) [inline build.gradle]: Why is Jacoco added right into the application itself?

File: build.gradle [hunk 5]
```diff
@@ -14,13 +15,21 @@ ktlint {
     enableExperimentalRules.set(true)
 }
 
+compileKotlin { kotlinOptions.jvmTarget = 1.8 }
+
 dependencies {
     testImplementation("org.junit.jupiter:junit-jupiter-engine:5.8.1")
     testImplementation("org.junit.jupiter:junit-jupiter-params:5.8.1")
-    implementation "org.jetbrains.kotlin:kotlin-stdlib"
+    implementation("org.jetbrains.kotlin:kotlin-stdlib")
```
Comment (TanVD) [inline build.gradle:24]: Use `implementation(kotlin("stdlib"))`

File: build.gradle [hunk 6]
```diff
@@ -14,13 +15,21 @@ ktlint {
     enableExperimentalRules.set(true)
 }
 
+compileKotlin { kotlinOptions.jvmTarget = 1.8 }
+
 dependencies {
     testImplementation("org.junit.jupiter:junit-jupiter-engine:5.8.1")
```
Comment (TanVD) [inline build.gradle:21]: No need in testRuntime dependency?

File: build.gradle [hunk 7]
```diff
@@ -1,6 +1,7 @@
 plugins {
     id 'org.jetbrains.kotlin.jvm' version '1.5.31'
     id("org.jlleitschuh.gradle.ktlint") version "10.0.0" apply true
+    id 'jacoco'
 }
 
 group 'org.example'
```
Comment (TanVD) [inline build.gradle]: It would be cool to have a bit more descriptive name of group

**Comment by TanVD** [comment]
For Gradle task -- 7

File: src/test/kotlin/ru/hse/smartUniversity/RouteFindingTest.kt (hunk 3 above)
Comment (sarnatk) [reply src/test/kotlin/ru/hse/smartUniversity/RouteFindingTest.kt:7]: I'd like to disagree here
"Factories" don't lessen the amount of cases
You can see, that each map in the parameters block has its own meaningful name, which represent one particular corner case
So "factories" basically help us to get rid of redundant code duplication (blah blah blah assert(actual, expected))
Moreover, if you run this code in IDE, it will show you that amount of tests _is_ actually 15, because what we did is just a kotlin syntactic sugar, and we really do test different cases by passing its maps as parameters to a templated testing function

File: build.gradle (hunk 4 above)
Comment (sarnatk) [reply build.gradle]: Removed

File: build.gradle (hunk 5 above)
Comment (sarnatk) [reply build.gradle:24]: Done

File: build.gradle (hunk 6 above)
Comment (sarnatk) [reply build.gradle:21]: Added

File: build.gradle (hunk 7 above)
Comment (sarnatk) [reply build.gradle]: Now it is cool

**Comment by sarnatk** [comment]
Also removed redundant "chmod +x gradlew" and refactored editorconfig

**Comment by sarnatk** [comment]
Build failed as "Could not find method kotlin() for arguments [stdlib]" so I returned to previous version

File: build.gradle (hunk 5 above)
Comment (sarnatk) [reply build.gradle:24]: But then removed as it wasn't working and also caused warnings in IDE
